lib/elixir/lib/module/types/descr.ex

```diff
@@ -1066,6 +1078,308 @@ defmodule Module.Types.Descr do
+          [{tag, fields, negs1 ++ negs2} | acc]
+        catch
+          :empty -> acc
+        end
+    end
+    |> case do
+      [] -> 0
+      acc -> acc
+    end
+  end
+
+  defp tuple_literal_intersection(tag1, elements1, tag2, elements2) do
+    n = length(elements1)
+    m = length(elements2)
+
+    cond do
+      (tag1 == :closed and n < m) or (tag2 == :closed and n > m) -> throw(:empty)
+      tag1 == :open and tag2 == :open -> {:open, zip_intersection(elements1, elements2, [])}
+      true -> {:closed, zip_intersection(elements1, elements2, [])}
+    end
+  end
+
+  # Intersects two lists of types, and _appends_ the extra elements to the result.
+  defp zip_intersection([], types2, acc), do: Enum.reverse(acc, types2)
+  defp zip_intersection(types1, [], acc), do: Enum.reverse(acc, types1)
+
+  defp zip_intersection([type1 | rest1], [type2 | rest2], acc) do
+    zip_intersection(rest1, rest2, [non_empty_intersection!(type1, type2) | acc])
+  end
+
+  defp tuple_difference(dnf1, dnf2) do
+    Enum.reduce(dnf2, dnf1, fn {tag2, elements2, negs2}, dnf1 ->
+      Enum.reduce(dnf1, [], fn {tag1, elements1, negs1}, acc ->
+        acc = [{tag1, elements1, [{tag2, elements2} | negs1]}] ++ acc
+
+        Enum.reduce(negs2, acc, fn {neg_tag2, neg_elements2}, acc ->
+          try do
+            {tag, fields} = tuple_literal_intersection(tag1, elements1, neg_tag2, neg_elements2)
+            [{tag, fields, negs1}] ++ acc
+          catch
+            :empty -> acc
+          end
+        end)
+      end)
+    end)
+    |> case do
+      [] -> 0
+      acc -> acc
+    end
+  end
+
+  defp tuple_union(left, right), do: left ++ right
+
+  defp tuple_to_quoted(dnf) do
+    dnf
+    |> tuple_normalize()
+    |> Enum.map(&tuple_each_to_quoted/1)
+    |> case do
+      [] -> []
+      dnf -> Enum.reduce(dnf, &{:or, [], [&2, &1]}) |> List.wrap()
+    end
+  end
+
+  defp tuple_each_to_quoted({tag, positive_map, negative_maps}) do
+    case negative_maps do
+      [] ->
+        tuple_literal_to_quoted({tag, positive_map})
+
+      _ ->
+        negative_maps
+        |> Enum.map(&tuple_literal_to_quoted/1)
+        |> Enum.reduce(&{:or, [], [&2, &1]})
+        |> Kernel.then(
+          &{:and, [], [tuple_literal_to_quoted({tag, positive_map}), {:not, [], [&1]}]}
+        )
+    end
+  end
+
+  defp tuple_literal_to_quoted({:closed, []}), do: {:{}, [], []}
+
+  defp tuple_literal_to_quoted({tag, elements}) do
+    case tag do
+      :closed -> {:{}, [], Enum.map(elements, &to_quoted/1)}
+      :open -> {:{}, [], Enum.map(elements, &to_quoted/1) ++ [{:..., [], nil}]}
+    end
+  end
+
+  # Pads a list of elements with term().
+  defp tuple_fill(elements, length) when length(elements) > length do
+    raise ArgumentError, "tuple_fill: elements are longer than the desired length"
+  end
+
+  defp tuple_fill(elements, desired_length) do
+    elements ++ List.duplicate(term(), desired_length - length(elements))
```

Comment:
Instead of computing length(elements) twice, do it once and check if `desired_length - length(elements)` is less than zero here?